lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,594 @@
+        if not relation_data:
+            logger.info("No relation data - Deferring")
+            return
+        if not self._relation_data_is_valid(relation_data):
+            logger.warning("Relation data did not pass JSON Schema validation - Deferring")
+            return
+        for certificate_request in relation_data.get("certificate_signing_requests", {}):
+            self.on.certificate_request.emit(
+                certificate_signing_request=certificate_request.get("certificate_signing_request"),
+                relation_id=event.relation.id,
+            )
+
+
+class TLSCertificatesRequires(Object):
+    """TLS certificates requirer class to be instantiated by TLS certificates requirers."""
+
+    on = CertificatesRequirerCharmEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relationship_name: str,
+    ):
+        """Generates/use private key and observes relation changed event.
+
+        Args:
+            charm: Charm object
+            relationship_name: Juju relation name
+        """
+        super().__init__(charm, relationship_name)
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_relation_changed
+        )
+        self.relationship_name = relationship_name
+        self.charm = charm
+
+    def request_certificate(
+        self,
+        private_key: bytes,
+        private_key_password: bytes,
+        common_name: str,
+        sans: list = None,
+    ) -> bytes:
+        """Request TLS certificate to provider charm.
+
+        Args:
+            private_key (bytes): Private key
+            private_key_password (bytes): Private key password
+            common_name (str): Common name.
+            sans (list): Subject Alternative Name
+
+        Returns:
+            bytes: The CSR passed in the relation data
+        """
+        logger.info("Received request to create certificate")
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            message = (
+                f"Relation {self.relationship_name} does not exist - "
+                f"The certificate request can't be completed"
+            )
+            logger.error(message)
+            raise RuntimeError(message)
+        relation_data = _load_relation_data(relation.data[self.model.unit])
+        csr = generate_csr(
+            private_key=private_key,
+            private_key_password=private_key_password,
+            subject=common_name,
+            sans=sans,
+        )
+        new_certificate_request = {"certificate_signing_request": csr.decode()}
+        if "certificate_signing_requests" in relation_data:
+            certificate_request_list = relation_data["certificate_signing_requests"]
+            if new_certificate_request in certificate_request_list:
+                logger.info("Request was already made - Doing nothing")
+                return csr
```

Comment:
we need to have mechanic to renew certificates

Reply:
From what I've seen from digicert and let's encrypt, a new CSR is created for each renewal so renewal is equivalent to simply creating a new certificate. For the tls-certificates interface that means no change. So for here in the library, I think it would be great to have an event triggered when certificates are about to expire.

Reply:
This is implemented now